FirebaseSessions/Sources/SessionInitiator.swift

```diff
@@ -22,8 +28,72 @@ import Foundation
 ///      and comes to the foreground.
 ///
 class SessionInitiator {
+  let sessionTimeout: TimeInterval = 30 * 60 // 30 minutes
+  let currentTime: () -> Date
+  var backgroundTime = Date.distantFuture
+  var initiateSessionStart: () -> Void = {}
+
+  init(dateProvider: @escaping () -> Date = Date.init) {
+    currentTime = dateProvider
+  }
+
   func beginListening(initiateSessionStart: @escaping () -> Void) {
-    // Only cold start is implemented right now
-    initiateSessionStart()
+    self.initiateSessionStart = initiateSessionStart
+    self.initiateSessionStart()
+
+    let notificationCenter = NotificationCenter.default
+    #if os(iOS) || os(tvOS)
+      notificationCenter.addObserver(
+        self,
+        selector: #selector(appBackgrounded),
+        name: UIApplication.didEnterBackgroundNotification,
+        object: nil
+      )
+      notificationCenter.addObserver(
+        self,
+        selector: #selector(appForegrounded),
+        name: UIApplication.didBecomeActiveNotification,
+        object: nil
+      )
+    #elseif os(OSX)
+      notificationCenter.addObserver(
+        self,
+        selector: #selector(appBackgrounded),
+        name: NSApplication.didResignActiveNotification,
+        object: nil
+      )
+      notificationCenter.addObserver(
+        self,
+        selector: #selector(appForegrounded),
+        name: NSApplication.didBecomeActiveNotification,
+        object: nil
+      )
+    #elseif os(watchOS)
+      if #available(watchOSApplicationExtension 7.0, *) {
```

Comment:
Can we add a comment here indicating the current situation? Eg. "watchOS versions less than 7.0 don't support lifecycle events"

Reply:
Done